src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1117,245 @@ private static string RegularExpressionFromFileSpec
+            }
+            else if (IsSpecialRegexCharacter(ch))
+            {
+                regex.Append('\\');
+                regex.Append(ch);
+            }
+            else
+            {
+                regex.Append(ch);
             }
+        }
 
-            /*
-             * Remaining call-outs not involving "**"
-             */
-            matchFileExpression.Replace("*.*", "<:anynonseparator:>");
-            matchFileExpression.Replace("*", "<:anynonseparator:>");
-            matchFileExpression.Replace("?", "<:singlecharacter:>");
+        private static bool IsSpecialRegexCharacter(char ch) =>
+            ch == '\\' || ch == '$' || ch == '(' || ch == ')' || ch == '+'
+            || ch == '.' || ch == '[' || ch == '^' || ch == '{' || ch == '|';
+
+        /*
+         * Iteratively skip three cases involving directory separators
+         * 
+         *  (1) \.\ -> \
+         *        This is an identity, so for example, these two are equivalent,
+         * 
+         *            dir1\.\dir2 == dir1\dir2
+         * 
+         *    (2) \\ -> \
+         *      Double directory separators are treated as a single directory separator,
+         *      so, for example, this is an identity:
+         * 
+         *          f:\dir1\\dir2 == f:\dir1\dir2
+         * 
+         *      The single exemption is for UNC path names, like this:
+         * 
+         *          \\server\share != \server\share
+         * 
+         *      This case is handled by isUncPath in
+         *      a prior step.
+         *
+         *  (3) **\** -> **
+         */
+        private static int IndexOfNextNonCollapsibleChar(string str, int startIndex)
+        {
+            if (startIndex > 0 && !FileUtilities.IsAnySlash(str[startIndex - 1]))
+            {
+                return startIndex;
+            }
+            int i = startIndex;
+            bool isNonCollapsibleCharFound = false;
 
-            /*
-             *  Escape all special characters defined for regular expresssions.
-             */
-            matchFileExpression.Replace("\\", "\\\\"); // Must be first.
-            matchFileExpression.Replace("$", "\\$");
-            matchFileExpression.Replace("(", "\\(");
-            matchFileExpression.Replace(")", "\\)");
-            matchFileExpression.Replace("*", "\\*");
-            matchFileExpression.Replace("+", "\\+");
-            matchFileExpression.Replace(".", "\\.");
-            matchFileExpression.Replace("[", "\\[");
-            matchFileExpression.Replace("?", "\\?");
-            matchFileExpression.Replace("^", "\\^");
-            matchFileExpression.Replace("{", "\\{");
-            matchFileExpression.Replace("|", "\\|");
+            while (!isNonCollapsibleCharFound && i < str.Length)
+            {
+                bool isSeparator = FileUtilities.IsAnySlash(str[i]);
+                bool isRelativeSeparator = i < str.Length - 1 && str[i] == '.' && FileUtilities.IsAnySlash(str[i + 1]);
```

Comment:
Shouldn't this check for `/./` instead of `./`? What if the path contains a directory ending in `.`?

Reply:
Prev slash is implicit thanks to the first check in the method